runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts

```diff
@@ -0,0 +1,71 @@
+import { Injectable } from '@nestjs/common';
+import { ProcessInstanceEventStatus, ProcessInstanceStatus } from 'runbotics-common';
+import { ProcessInstanceEventEntity } from 'src/database/process-instance-event/process-instance-event.entity';
+import { ProcessInstanceEventService } from 'src/database/process-instance-event/process-instance-event.service';
+import { ProcessInstanceService } from 'src/database/process-instance/process-instance.service';
+import { Logger } from 'src/utils/logger';
+import { Connection } from 'typeorm';
+
+@Injectable()
+export class BotLifecycleService {
+    constructor(
+        private readonly processInstanceService: ProcessInstanceService,
+        private readonly processInstanceEventService: ProcessInstanceEventService,
+        private readonly connection: Connection,
+    ) { }
+    
+    private readonly logger = new Logger(BotLifecycleService.name);
+
+    async handleInterruptedProcessInstanceExecution(botId: number): Promise<void> {
+        const disconnectedBotProcessInstances = await this.processInstanceService.findAllByBotId(botId);
+        disconnectedBotProcessInstances.forEach(async (processInstance) => {
+            if (
+                processInstance.status === ProcessInstanceStatus.INITIALIZING ||
+                processInstance.status === ProcessInstanceStatus.IN_PROGRESS
+            ) {
+                const newProcessInstance = { 
+                    ...processInstance,
+                    status: ProcessInstanceStatus.ERRORED,
+                    error: 'Bot has been shut down',
+                };
+
+                await this.processInstanceService.save(newProcessInstance)
+                    .then(() => {
+                        this.logger.log(
+                            `<= Success: process-instance (${
+                                newProcessInstance.id
+                            }) updated after bot disconnection | status: ${
+                                newProcessInstance.status
+                            }, error message: "${
+                                newProcessInstance.error
+                            }"`
+                        );
+                    })
+                    .catch((err) => {
+                        this.logger.log(`<= Error: process-instance (${newProcessInstance.id}) update after bot disconnection failed | Error: ${err}`);
+                    });
+
+                this.handleInterruptedProcessInstanceEvent(processInstance.id);
+            }
+        });
+    }
+
+    async handleInterruptedProcessInstanceEvent (processInstanceId: string): Promise<void> {
+        const activeProcessInstanceEvents = await this.processInstanceEventService.findAllActiveByProcessInstanceId(processInstanceId);
+
+        if (activeProcessInstanceEvents.length > 0) {
```

Comment:
make it guard if, reduce nesting if possible

Reply:
roger that, done